### PR TITLE
Fixing calling `.visualize()` with `filename=None`

### LINKED ISSUE #7685: `dask.visualize` does not work with `filename=None`
According to the documentation for [`dask.visualize`](https://docs.dask.org/en/latest/api.html#dask.visualize),

> **If filename is None, no file will be written**, and we communicate with dot using only pipes.

Unfortunately, it seems this actually throws an error!

```
/python3.8/site-packages/dask/dot.py in <genexpr>(.0)
    274 def graphviz_to_file(g, filename, format):
    275     fmts = [".png", ".pdf", ".dot", ".svg", ".jpeg", ".jpg"]
--> 276     if format is None and any(filename.lower().endswith(fmt) for fmt in fmts):
    277         filename, format = os.path.splitext(filename)
    278         format = format[1:].lower()

AttributeError: 'NoneType' object has no attribute 'lower'
```

Based on the documentation, it sounds like `filename=None` is meant to be supported, but [based on the tests for `dask.visualize`](https://github.com/dask/dask/blob/main/dask/tests/test_base.py#L936-L955), this case was never actually tested for and this error slipped under the radar.

### PR DESCRIPTION
**Pull request by freyam**
- [x] Closes #7685 
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask` / `isort dask`

## The Issue
The issue arised when the `format` and the `filename` arguments both were `None`. So, 

https://github.com/dask/dask/blob/1a6a85258bccb14fe37bb54c211beeeffd46cd55/dask/dot.py#L276-L278

would result in an error as it's trying to call the `.lower()` function on a `NoneType` filename.

## The Fix
To fix this, I added an extra condition before it reaches the line of error to make sure `format` is not `None` by making it equal to the default ("png").

## The Demo
image

### CODE REVIEW

**Comment by freyam** [comment]
@martindurant @GenevieveBuckley 

The issue arised when the `format` and the `filename` arguments both were `None`. So, 

https://github.com/dask/dask/blob/1a6a85258bccb14fe37bb54c211beeeffd46cd55/dask/dot.py#L276-L278

would result in an error as it's trying to run `.lower()` on a `NoneType`.

To fix this, I added an extra condition before it reaches the line of error to make sure `format` is not `None` by making it equal to the default ("png").

image

**Comment by freyam** [comment]
What kind of tests can I add for this? Not very sure since there is no file being generated!
https://github.com/dask/dask/blob/1a6a85258bccb14fe37bb54c211beeeffd46cd55/dask/tests/test_base.py#L936-L955
IIn this case, what should I check for?

EDIT: surprisingly, this fails many test cases. Will have a look and commit the fixes.

File: dask/dot.py
Comment (martindurant) [inline dask/dot.py:283]: Cat it happen that `format` is still `None` here, e.g., a filename was given but didn't match any of the known extensions?

File: dask/dot.py
Comment (martindurant) [inline dask/dot.py:297]: You could use `monkeypatch` to temporarily change `display_cls`, so that it has an effect you can check from the test function.

Do you know how to do this?

**Review by martindurant — COMMENTED** [review]
> What kind of tests can I add for this

At least, call the function with the arguments that previously caused an exception, and you should see that there is no exception. I agree that, since this is not run within jupyter, you cannot know if it succeeds.

File: dask/dot.py
Comment (freyam) [reply dask/dot.py:283]: Oh yes! I shouldn't have deleted the previous check for that. This was the issue! The failed tests pass now :tada: :heavy_check_mark:

File: dask/dot.py
Comment (freyam) [reply dask/dot.py:297]: No, never heard of `monkeypatch`. Can you point me to relevant resources?

Comment (freyam) [reply dask/dot.py:297]: I went through [this](https://docs.pytest.org/en/6.2.x/monkeypatch.html) and this is exactly what I was looking for! Wow :100:

Comment (martindurant) [reply dask/dot.py:297]: Ping me if you cannot quickly figure it out, I can show you how I would do this.

File: dask/dot.py
Comment (GenevieveBuckley) [inline dask/dot.py:280]: Should these two if statements be logically grouped with each other?

```python
if format is None and filename is None:
    pass
elif format is None and any(filename.lower().endswith(fmt) for fmt in fmts):
    # do the thing...
```

File: dask/tests/test_base.py
Comment (GenevieveBuckley) [inline dask/tests/test_base.py:957]: It's confusing to work out what is actually being tested just from reading the new test code you've added. To address this, I'd probably recommend splitting it up into it's own separate test and call that something like `test_visualize_no_filename`. Then the name of the test can do the work of describing what the test is for.

Whether something should be a new test, or part of an existing test is a bit of a borderline judgement call, and other people might disagree with me. Ultimately, I think it'lll save time for future developers because it's a clearer way to mark "you can ignore this bit" if they're not looking into the specific case where `filename=None`.

File: dask/tests/test_base.py
Comment (GenevieveBuckley) [inline dask/tests/test_base.py:960]: It's possible I'm a bit confused here, but isn't the mocked behaviour for the function the same as the regular behaviour?
If that's true, do we really need to add the extra complexity of mocking? It might be a lot simpler to just add `visualize(x, filename=None)` as a line in `test_visualize` and a comment indicating that the test will fail if an error is raised by that line of code. 

(Note: that's the pytest default, a test will fail if an error is raised anywhere in it, but a comment is helpful for developers who are generally looking for assert statements to tell them what the test is about)

Comment (GenevieveBuckley) [reply dask/tests/test_base.py:960]: @martindurant have I misunderstood something about the mocking here (entirely possible) or is it a bit overkill?

**Comment by GenevieveBuckley** [comment]
This is nice work @freyam, thank you for making this!

File: dask/dot.py
Comment (freyam) [reply dask/dot.py:280]: Oh yes, this makes more sense. Updated :+1:

File: dask/tests/test_base.py
Comment (freyam) [reply dask/tests/test_base.py:957]: Yes, this seems to be a better option.

```py
@pytest.mark.skipif("not da")
@pytest.mark.skipif(
    sys.flags.optimize, reason="graphviz exception with Python -OO flag"
)
def test_visualize_no_filename(monkeypatch):
    pytest.importorskip("graphviz")
    pytest.importorskip("dask.dot")

    x = da.arange(10)

    test_list = []
    _mock_get_display_cls = dask.dot._get_display_cls

    def mock_fx(format):
        test_list.append(True)
        return _mock_get_display_cls(format)

    monkeypatch.setattr(dask.dot, "_get_display_cls", mock_fx)
    visualize(x, filename=None)
    assert True in test_list
```
  
I can add this (the test :heavy_check_mark:)
I will wait for the final decision on using monkeypatch to mock behavior and then commit accordingly!

File: dask/tests/test_base.py
Comment (martindurant) [reply dask/tests/test_base.py:960]: It is the same behaviour, but this slightly odd construct makes sure that the function actually got called (as opposed to None being some kind of no-op). Could be removed for style's sake.

File: dask/tests/test_base.py
Comment (martindurant) [reply dask/tests/test_base.py:957]: I have no opinion on one test versus two

**Comment by freyam** [comment]
I have pushed a separate function for the `no_filename` test and I have kept the monkeypatch mock behavior.
Is this fine? @martindurant @GenevieveBuckley 
or should I just include a one-line test and commend it accordingly?


```py
# To check if visualize function works when filename is None
# Test will fail if the function raises an error
visualize(x, filename=None)
```

File: dask/dot.py
Comment (martindurant) [inline dask/dot.py:279]: Actually, these three lines could be condensed, either
```
if format is None and filename is not None and any(filename.lower().endswith(fmt) for fmt in fmts):
```
or
```
if format is None and any((filename or "").lower().endswith(fmt) for fmt in fmts):
```
The second is shorter, the first is clearer.

File: dask/tests/test_base.py
Comment (martindurant) [inline dask/tests/test_base.py:971]: As @GenevieveBuckley  said, a comment here for what we're doing (this patch ensures that display_cls got called successfully) would be good.

**Review by martindurant — COMMENTED** [review]
I am happy with it, but I still have a couple of comments.
@GenevieveBuckley , you can decide on final style here, I don't really mind.

File: dask/dot.py
Comment (freyam) [reply dask/dot.py:279]: I added the first one! Even though it's longer, using `black` formatting, it looks very clean.

```py
if (
    format is None
    and filename is not None
    and any(filename.lower().endswith(fmt) for fmt in fmts)
):
    filename, format = os.path.splitext(filename)
    format = format[1:].lower()
```

File: dask/tests/test_base.py
Comment (freyam) [reply dask/tests/test_base.py:971]: Done!

**Comment by freyam** [comment]
Oops! I have been working on the `main` branch of my forked repository. So, while pulling changes, it got considered as a commit towards this PR.

I am not sure how I overlooked this, I will make sure to use a new branch from next time!

For now, you can consider the second last commit. That passes all the tests.

https://github.com/dask/dask/pull/7740/commits/9bdcbaa8e1323ea0b83bee5740a94628ceff1624

File: dask/tests/test_base.py
Comment (GenevieveBuckley) [inline dask/tests/test_base.py:968]: For clarity, you might like to rename `test_list` to something more meaningful. Maybe something like `was_mock_function_called`?

**Comment by GenevieveBuckley** [comment]
> LGTM, but someone else should pull the trigger if they are also happy.

ping @dask/maintenance

**Comment by mrocklin** [comment]
A few requests:

1.  In the future, can I ask for a more informative PR title?  This is for a few reasons:
    - Very few people know all of the issue numbers, and so saying "Fixed #XXXX" doesn't help reviewers know what's going on
    - These titles become the title of the commit
2. Can I ask for an informative comment when opening the PR.  This might briefly describe the problem, and then point to the original issue for greater details.  It should then also describe the solution taken so that reviewers can get a sense for what you're trying to do without looking through your code
3. I would prefer not to use monkeypatch if we can avoid it.  At a quick glance at the test I don't know what's going on. I'm sure that it's correct, but it does require a bit more from reviewers to understand.  It's also not typically used in this codebase.

**Comment by martindurant** [comment]
OK, from two counts against, lets remove monkeypatch after all

**Comment by freyam** [comment]
@mrocklin Yes, I will definitely take care of that. Thank you so much for pointing this out! :100:

**Comment by freyam** [comment]
I just made a simple one-liner test in the already existing `test_visualize()`.

Also, once this PR is merged, I will edit the topmost comment with relevant information about the issue and how I solved it.

**Comment by martindurant** [comment]
(I think Matt is asking that you should ensure your title is descriptive before merger)

**Comment by freyam** [comment]
@martindurant Oh yes! Changed :)

**Comment by martindurant** [comment]
I took the liberty of simplifying it a little

**Comment by GenevieveBuckley** [comment]
> Also, once this PR is merged, I will edit the topmost comment with relevant information about the issue and how I solved it.

You don't need to wait until after it's merged, you can go ahead and edit the top description now.

**Comment by GenevieveBuckley** [comment]
The pytest output is too verbose, according to GitHub actions which has truncated it. So I *think*, but do not know for sure that failing tests are mostly in dask bag and unrelated to your PR. All the tests pass locally for me, so I'm not entirely sure why our CI is unhappy.

You can re-trigger the CI by adding an [empty git commit](https://coderwall.com/p/vkdekq/git-commit-allow-empty): git commit --allow-empty -m "[EMPTY] Re-trigger CI"``

**Comment by jrbourbeau** [comment]
Just opened https://github.com/dask/dask/pull/7771 to run CI against the current `main` branch. That should help us determine if there are CI failures unrelated to the changes here

**Comment by freyam** [comment]
@GenevieveBuckley  Yes, I have already edited the topmost comment to follow the format -> The Issue, The Fix, and The Demo.

Also, I have just submitted an empty commit. Let's see how it performs.

**Comment by freyam** [comment]
Looks like @jrbourbeau's suspicion was right. I will wait till #7771 is fixed before committing.

**Comment by freyam** [comment]
Fingers Crosses! :crossed_fingers:

**Comment by freyam** [comment]
> :red_circle: **Error: The process '/usr/share/miniconda3/condabin/mamba' failed with exit code 1**

**Comment by GenevieveBuckley** [comment]
Hm, I'm not sure what might have caused that mamba error. You can try the empty commit trick to check if it's just a momentary bit of weirdness (if you see the same thing twice, that might indicate a problem)

**Comment by freyam** [comment]
YEAH! Wow, what a roller-coaster ride this issue was! I'm glad it all worked out in the end: :rocket: Learned a lot from this issue. I will try not to repeat any of the mistakes I made this time :heavy_check_mark:

**Review by GenevieveBuckley — APPROVED** [review]
:tada: Good work @freyam!

I'm pleased to see alll the tests are passing now too. Thanks @jrbourbeau for tracking down what was going on.

**Comment by martindurant** [comment]
Hurray!

**Pull request merged**
